core-bundle/src/Resources/contao/library/Contao/Search.php

```diff
@@ -181,27 +181,20 @@ public static function indexPage($arrData)
 								->execute($arrSet['checksum'], $arrSet['pid']);
 
 		// Update the URL if the new URL is shorter or the current URL is not canonical
-		if ($objIndex->numRows && $objIndex->url != $arrSet['url'])
+		if ($objIndex->numRows)
 		{
-			if (strpos($arrSet['url'], '?') === false && strpos($objIndex->url, '?') !== false)
+			// The new URL is more canonical (shorter and/or less fragments)
+			if (self::compareUrls($arrSet['url'], $objIndex->url) < 0)
 			{
-				// The new URL is more canonical (no query string)
-				$objDatabase->prepare("DELETE FROM tl_search WHERE id=?")
-							->execute($objIndex->id);
+				$objDatabase->prepare("DELETE FROM tl_search WHERE url=?")
```

Comment:
Why are we still deleting from `tl_search`? And what about the child records in `tl_search_index`?

Reply:
If the new URL is already present with a different checksum. This should not happen under normal circumstances, but to be sure that we don’t end up with duplicate URLs I think it makes sense to delete it here.

Reply:
And what about the child records in `tl_search_index`? We need to delete them, too, IMHO.

Reply:
Done in b502ff17767391c6801adc2177a18b05176c96fe